### PR TITLE
[AMD][Bugfix][Quantization] Honor fused-name match in is_layer_skipped

### PR DESCRIPTION
**Pull request by ZiguanWang**
## Purpose

PR #41892 (https://github.com/vllm-project/vllm/pull/41892) introduced a bug that breaks Step-3.5-Flash-FP8 on vLLM v0.22.0+: every prompt degenerates to a stream of `<｜begin▁of▁sentence｜>` tokens regardless of input. Reproduced on ROCm vLLM 0.22.0+rocm722, TP=4, MI308X with the official `stepfun-ai/Step-3.5-Flash-FP8` checkpoint.

```
[DEGENERATED] finish=length
  prompt    : 'The capital of France is'
  completion: '<｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜>'

[DEGENERATED] finish=length
  prompt    : '1 + 1 ='
  completion: '<｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜>'

[DEGENERATED] finish=length
  prompt    : 'Hello, my name is'
  completion: '<｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜>'

[DEGENERATED] finish=length
  prompt    : 'Once upon a time,'
  completion: '<｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜><｜begin▁of▁sentence｜>'

```

Root cause: #41892 registered a `packed_modules_mapping = {qkv_proj: [q_proj,k_proj,v_proj], gate_up_proj: [gate_proj,up_proj]}` on `Step3p5ForCausalLM` so that Quark INT8 exports (whose `modules_to_not_convert` lists unfused per-shard names) are handled correctly，https://huggingface.co/nameistoken/Step-3.5-Flash-Quark-W8A8-INT8/blob/main/config.json.
 However, the block-FP8 Step-3.5-Flash-FP8 checkpoint lists **fused** names directly in `modules_to_not_convert` (e.g. `model.layers.0.self_attn.qkv_proj`, `model.layers.0.mlp.gate_up_proj`). With the new mapping in place, `is_layer_skipped` unconditionally expands `qkv_proj` to its unfused shards and looks those up in `ignored_layers`. The unfused names are not present, so attention/MLP layers that should stay in BF16 are routed through `Fp8LinearMethod` and produce corrupted logits.
https://huggingface.co/stepfun-ai/Step-3.5-Flash-FP8/blob/main/config.json

Fix: in `is_layer_skipped`, when the projection name is in `fused_mapping` AND the fused prefix itself already matches an entry in `ignored_layers`, return skipped directly. Only fall back to the per-shard expansion when the fused name does not match. This keeps the Quark INT8 behavior from #41892 intact while restoring correct behavior for FP8/other exports that list fused names. One file changed, +9/-1.

## Test Plan

Verification on ROCm MI308X (TP=4) with Step-3.5-Flash-FP8:
  * Without this commit: `test_raw_completion.py` produces wrong output -- every prompt degenerates to repeated `<｜begin▁of▁sentence｜>` tokens.
  * With this commit: `test_raw_completion.py` output is normal (coherent completions on 4/4 sanity prompts), and gsm8k eval also runs normally.

## Test Result

```
[OK] finish=length
  prompt    : 'The capital of France is'
  completion: ' Paris. The capital of Germany is Berlin. The capital of Italy is Rome. The capital of Spain'

[OK] finish=length
  prompt    : '1 + 1 ='
  completion: ' 3: The Myth of the Two-Party System\n\nby\n\nMichael J. Malbin\n\n'

[OK] finish=length
  prompt    : 'Hello, my name is'
  completion: ' Dr. David F. Dinges. I am a Professor in the Department of Psychiatry at the University'

[OK] finish=length
  prompt    : 'Once upon a time,'
  completion: ' in a small village nestled in the heart of a dense forest, there lived a young girl named Lily'
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by ZiguanWang** [comment]
@tjtanaa @JoursBleu @BowenBao Can you take a look? Thanks!

**Comment by JoursBleu** [comment]
LGTM — clean minimal fix, correctly restores fused-name matching for FP8 checkpoints.

**Review by BowenBao — APPROVED** [review]
LGTM

**Review by tjtanaa — APPROVED** [review]

File: vllm/model_executor/layers/quantization/utils/quant_utils.py [hunk 1]
```diff
@@ -513,7 +513,15 @@ def substr_match(prefix: str, ignored_layers: list[str]) -> bool:
...
     # from the fused version to unfused + check to make sure that
     # each shard of the fused layer has the same scheme.
-    if proj_name in fused_mapping:
+    #
+    # Some checkpoints (e.g. block-FP8 Step-3.5-Flash) already list the
+    # fused name (e.g. ``self_attn.qkv_proj``) directly in
+    # ``modules_to_not_convert``. Honor that fused-name match first so
+    # those layers are still correctly skipped even when a
+    # ``packed_modules_mapping`` is registered on the model.
+    if proj_name in fused_mapping and match_func(prefix, ignored_layers):
```
Comment (okorzh-amd) [inline vllm/model_executor/layers/quantization/utils/quant_utils.py:529]: please add a unit test so the regression can be caught earlier next time.

Comment (ZiguanWang) [reply vllm/model_executor/layers/quantization/utils/quant_utils.py:529]: Should we add unit tests for Step3.5-Flash-FP8, or add unit tests for the logic introduced in this part?

Comment (okorzh-amd) [reply vllm/model_executor/layers/quantization/utils/quant_utils.py:529]: this `add unit tests for the logic introduced in this part`.

Comment (ZiguanWang) [reply vllm/model_executor/layers/quantization/utils/quant_utils.py:529]: added, could you help review it? thanks

File: tests/quantization/test_is_layer_skipped.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+"""Unit tests for ``is_layer_skipped`` fused-name handling."""
```
Comment (tjtanaa) [inline tests/quantization/test_is_layer_skipped.py]: @hmellor can you advise us where should we put the tests?

Comment (hmellor) [reply tests/quantization/test_is_layer_skipped.py]: These should likely belong in `vllm/tests/quantization/test_quark.py` which contains tests for model setup and weight loading. This is exactly the area that this PR touches.

Comment (tjtanaa) [reply tests/quantization/test_is_layer_skipped.py]: @ZiguanWang Let's move the test to the correct file.

File: vllm/model_executor/layers/quantization/utils/quant_utils.py (hunk 1 above)
Comment (okorzh-amd) [reply vllm/model_executor/layers/quantization/utils/quant_utils.py:529]: Looks good. thanks

File: tests/quantization/test_is_layer_skipped.py (hunk 2 above)
Comment (ZiguanWang) [reply tests/quantization/test_is_layer_skipped.py]: Sure, Done

Comment (ZiguanWang) [reply tests/quantization/test_is_layer_skipped.py]: @hmellor @tjtanaa can you help merge the PR, thanks

**Comment by ZiguanWang** [comment]
@hmellor @tjtanaa @AndreasKaratzas Can you help merge this PR?
